File: doma-core/src/main/java/org/seasar/doma/jdbc/criteria/command/MappedObjectProvider.java
package org.seasar.doma.jdbc.criteria.command;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.function.Function;
import org.seasar.doma.def.PropertyDef;
import org.seasar.doma.internal.jdbc.command.AbstractObjectProvider;
import org.seasar.doma.jdbc.JdbcMappingVisitor;
import org.seasar.doma.jdbc.criteria.statement.Row;
import org.seasar.doma.jdbc.entity.Property;
import org.seasar.doma.jdbc.query.Query;

public class MappedObjectProvider<RESULT> extends AbstractObjectProvider<RESULT> {
  private final Function<Row, RESULT> mapper;
  private final JdbcMappingVisitor jdbcMappingVisitor;
  private final Map<PropertyDef<?>, Integer> indexMap;

  public MappedObjectProvider(
      Query query, List<PropertyDef<?>> propertyDefs, Function<Row, RESULT> mapper) {
    Objects.requireNonNull(query);
    Objects.requireNonNull(propertyDefs);
    Objects.requireNonNull(mapper);
    this.mapper = mapper;
    this.jdbcMappingVisitor = query.getConfig().getDialect().getJdbcMappingVisitor();
    indexMap = new HashMap<>(propertyDefs.size());
    int index = 1;
    for (PropertyDef<?> propertyDef : propertyDefs) {
      indexMap.put(propertyDef, index++);
    }
  }

  @Override
  @SuppressWarnings("unchecked")
  public RESULT get(ResultSet resultSet) throws SQLException {
    Row row =
        new Row() {
          @Override
          public <PROPERTY> PROPERTY get(PropertyDef<PROPERTY> propertyDef) {
            Integer index = indexMap.get(propertyDef);
            if (index == null) {
              throw new IllegalArgumentException(
                  "The propertyDef is unknown. " + propertyDef.getName());
            }
            Property<?, ?> property = propertyDef.asType().createProperty();
            try {
              fetch(resultSet, property, index, jdbcMappingVisitor);
            } catch (SQLException e) {
              throw new UncheckedSQLException(e);
            }
            return (PROPERTY) property.get();
          }
        };
    try {
      return mapper.apply(row);
    } catch (UncheckedSQLException e) {
      throw e.getCause();
    }
  }
}

File: doma-criteria/src/main/kotlin/org/seasar/doma/criteria/command/MappedObjectProvider.kt
package org.seasar.doma.criteria.command

import java.sql.ResultSet
import org.seasar.doma.criteria.declaration.Row
import org.seasar.doma.def.PropertyDef
import org.seasar.doma.internal.jdbc.command.AbstractObjectProvider
import org.seasar.doma.jdbc.query.Query

class MappedObjectProvider<RESULT>(
    query: Query,
    propDefs: List<PropertyDef<*>>,
    private val mapper: (Row) -> RESULT
) : AbstractObjectProvider<RESULT>() {

    private val jdbcMappingVisitor = query.config.dialect.jdbcMappingVisitor

    private val indexMap = propDefs.mapIndexed { index, propDef ->
        propDef to index + 1
    }.toMap()

    override fun get(resultSet: ResultSet?): RESULT {
        val row = object : Row {
            override fun <PROPERTY> get(propDef: PropertyDef<PROPERTY>): PROPERTY? {
                val index = indexMap[propDef]
                require(index != null) {
                    "The propDef is unknown. " +
                        "Ensure that you specified it to the select function."
                }
                val prop = propDef.asType().createProperty()
                fetch(resultSet, prop, index, jdbcMappingVisitor)
                @Suppress("UNCHECKED_CAST")
                return prop.get() as PROPERTY?
            }
        }
        return mapper(row)
    }
}
